File: src/consume-number.js
/**
 * @fileoverview
 * Consumes a JSON number from JSONC input.
 */

import { isDigit } from './is-digit.js';

/**
 * @typedef {import('./tokenize-jsonc.js').Token} Token
 * @typedef {import('./tokenize-jsonc.js').TokenizeError} TokenizeError
 */

/**
 * Result from consumeNumber.
 * @typedef {Object} ConsumeResult
 * @property {number} column - Updated column position
 * @property {number} line - Line number (unchanged for numbers)
 * @property {number} pos - Updated position in the string
 * @property {Token} token - The consumed number token
 */

/**
 * Consumes a JSON number.
 * @param {string} jsonc - The input string
 * @param {number} startPos - Starting position
 * @param {number} len - Length of input
 * @param {number} line - Current line number
 * @param {number} startColumn - Starting column
 * @param {TokenizeError[]} errors - Errors array to append to
 * @returns {ConsumeResult}
 */
export function consumeNumber(jsonc, startPos, len, line, startColumn, errors) {
    let pos = startPos;
    let column = startColumn;
    let value = '';

    // Optional leading minus.
    if (pos < len && jsonc[pos] === '-') {
        value += '-';
        pos += 1;
        column += 1;
    }

    // Integer part.
    if (pos < len && jsonc[pos] === '0') {
        // Leading zero must not be followed by another digit.
        value += '0';
        pos += 1;
        column += 1;
    } else if (pos < len && isDigit(jsonc[pos])) {
        // Non-zero digit followed by optional digits.
        while (pos < len && isDigit(jsonc[pos])) {
            value += jsonc[pos];
            pos += 1;
            column += 1;
        }
    } else {
        // No digit after minus.
        errors.push({
            category: 'syntax',
            column: startColumn,
            line,
            message: 'Invalid number: expected digit',
        });
        return {
            column,
            line,
            pos,
            token: { category: 'number', column: startColumn, line, value },
        };
    }

    // Optional fractional part.
    if (pos < len && jsonc[pos] === '.') {
        value += '.';
        pos += 1;
        column += 1;
        // Must have at least one digit after decimal point.
        if (pos >= len || !isDigit(jsonc[pos])) {
            errors.push({
                category: 'syntax',
                column,
                line,
                message: 'Invalid number: expected digit after decimal point',
            });
        } else {
            while (pos < len && isDigit(jsonc[pos])) {
                value += jsonc[pos];
                pos += 1;
                column += 1;
            }
        }
    }

    // Optional exponent part.
    if (pos < len && (jsonc[pos] === 'e' || jsonc[pos] === 'E')) {
        value += jsonc[pos];
        pos += 1;
        column += 1;
        // Optional sign.
        if (pos < len && (jsonc[pos] === '+' || jsonc[pos] === '-')) {
            value += jsonc[pos];
            pos += 1;
            column += 1;
        }
        // Must have at least one digit.
        if (pos >= len || !isDigit(jsonc[pos])) {
            errors.push({
                category: 'syntax',
                column,
                line,
                message: 'Invalid number: expected digit in exponent',
            });
        } else {
            while (pos < len && isDigit(jsonc[pos])) {
                value += jsonc[pos];
                pos += 1;
                column += 1;
            }
        }
    }

    return {
        column,
        line,
        pos,
        token: { category: 'number', column: startColumn, line, value },
    };
}

File: wasm/jsonc_lexer/src/consume_number.rs
//! Consumes a JSON number from JSONC input.

use crate::is_digit::is_digit;
use crate::types::{ConsumeResult, Token, TokenizeError};

/// Consumes a JSON number.
pub fn consume_number(
    bytes: &[u8],
    start_pos: usize,
    len: usize,
    line: u32,
    start_column: u32,
    errors: &mut Vec<TokenizeError>,
) -> ConsumeResult {
    let mut pos = start_pos;
    let mut column = start_column;
    let mut value = Vec::new();

    // Optional leading minus.
    if pos < len && bytes[pos] == b'-' {
        value.push(b'-');
        pos += 1;
        column += 1;
    }

    // Integer part.
    if pos < len && bytes[pos] == b'0' {
        // Leading zero must not be followed by another digit.
        value.push(b'0');
        pos += 1;
        column += 1;
    } else if pos < len && is_digit(bytes[pos]) {
        // Non-zero digit followed by optional digits.
        while pos < len && is_digit(bytes[pos]) {
            value.push(bytes[pos]);
            pos += 1;
            column += 1;
        }
    } else {
        // No digit after minus.
        errors.push(TokenizeError {
            category: "syntax".to_string(),
            column: start_column,
            line,
            message: "Invalid number: expected digit".to_string(),
        });
        return ConsumeResult {
            column,
            line,
            pos,
            token: Token {
                category: "number".to_string(),
                column: start_column,
                line,
                value: String::from_utf8(value).unwrap_or_default(),
            },
        };
    }

    // Optional fractional part.
    if pos < len && bytes[pos] == b'.' {
        value.push(b'.');
        pos += 1;
        column += 1;
        // Must have at least one digit after decimal point.
        if pos >= len || !is_digit(bytes[pos]) {
            errors.push(TokenizeError {
                category: "syntax".to_string(),
                column,
                line,
                message: "Invalid number: expected digit after decimal point".to_string(),
            });
        } else {
            while pos < len && is_digit(bytes[pos]) {
                value.push(bytes[pos]);
                pos += 1;
                column += 1;
            }
        }
    }

    // Optional exponent part.
    if pos < len && (bytes[pos] == b'e' || bytes[pos] == b'E') {
        value.push(bytes[pos]);
        pos += 1;
        column += 1;
        // Optional sign.
        if pos < len && (bytes[pos] == b'+' || bytes[pos] == b'-') {
            value.push(bytes[pos]);
            pos += 1;
            column += 1;
        }
        // Must have at least one digit.
        if pos >= len || !is_digit(bytes[pos]) {
            errors.push(TokenizeError {
                category: "syntax".to_string(),
                column,
                line,
                message: "Invalid number: expected digit in exponent".to_string(),
            });
        } else {
            while pos < len && is_digit(bytes[pos]) {
                value.push(bytes[pos]);
                pos += 1;
                column += 1;
            }
        }
    }

    ConsumeResult {
        column,
        line,
        pos,
        token: Token {
            category: "number".to_string(),
            column: start_column,
            line,
            value: String::from_utf8(value).expect("number should be valid UTF-8"),
        },
    }
}
